bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java

```diff
@@ -0,0 +1,225 @@
+
+    private static final long serialVersionUID = 8828949721938234629L;
+    private static final BigDecimal HUNDRED = BigDecimal.valueOf(100);
+
+    // Regular expression to split unit from value
+    private static final String UNIT_PATTERN = "(?<=\\d)\\s*(?=[a-zA-Z°µ])";
+
+    protected Quantity<?> quantity;
+
+    public QuantityType(String value) {
+        if (value != null) {
+            String[] constituents = value.split(UNIT_PATTERN);
+
+            // getQuantity needs a space between numeric value and unit
+            String formatted = String.join(" ", constituents);
+            quantity = Quantities.getQuantity(formatted);
+        } else {
+            throw new IllegalArgumentException("Constructor argument must not be null");
+        }
+    }
+
+    public QuantityType(double value, Unit<?> unit) {
+        quantity = Quantities.getQuantity(value, unit);
+    }
+
+    public static QuantityType valueOf(double value, Unit<?> unit) {
+        return new QuantityType(value, unit);
+    }
+
+    @Override
+    public String toString() {
+        return toFullString();
+    }
+
+    public static QuantityType valueOf(String value) {
+        return new QuantityType(value);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (!(obj instanceof QuantityType)) {
+            return false;
+        }
+        QuantityType other = (QuantityType) obj;
+        if (quantity == null) {
+            if (other.quantity != null) {
+                return false;
+            }
+        } else if (quantity.getUnit().getDimension() != other.quantity.getUnit().getDimension()) {
+            return false;
+        } else if (compareTo(other) != 0) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int compareTo(QuantityType o) {
+        if (quantity.getUnit().isCompatible(o.quantity.getUnit())) {
+            QuantityType v1 = this.toUnit(getUnit().getSystemUnit());
+            QuantityType v2 = o.toUnit(o.getUnit().getSystemUnit());
+            return Double.compare(v1.doubleValue(), v2.doubleValue());
+        } else {
+            throw new IllegalArgumentException("Can not compare incompatible units");
+        }
+    }
+
+    public Unit<?> getUnit() {
+        return quantity.getUnit();
+    }
+
+    public Dimension getDimension() {
+        return getUnit().getDimension();
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public QuantityType toUnit(Unit targetUnit) {
+        try {
+            if (targetUnit != getUnit()) {
+                UnitConverter uc = getUnit().getConverterTo(targetUnit);
+                Quantity result = Quantities.getQuantity(uc.convert(quantity.getValue()), targetUnit);
+
+                return new QuantityType(result.getValue().doubleValue(), targetUnit);
+            } else {
```

Comment:
The `else` is unnecessary here. You can even move the `return this`to the end of the method which is then the default case when the units dont match.